#polars is used for data loading and manipulation.
#elfen is used for linguistic feature extraction

In [ ]:
#Step 1: Import Required Libraries
import polars as pl
import elfen

#Load the text dataset into a DataFrame.
#Verify the dataset structure before analysis.

In [3]:
# Step 2: Load the Dataset
df = pl.read_csv("Random_English_Sentences.csv")

print(df.head())
print(df.shape)

shape: (5, 2)
┌─────┬─────────────────────────────────┐
│ id  ┆ text                            │
│ --- ┆ ---                             │
│ i64 ┆ str                             │
╞═════╪═════════════════════════════════╡
│ 1   ┆ I didn't know what to wear to … │
│ 2   ┆ There used to be a big cherry … │
│ 3   ┆ Sometimes, books and movies ma… │
│ 4   ┆ My bandaid wasn't sticky anymo… │
│ 5   ┆ He only had one drink but he's… │
└─────┴─────────────────────────────────┘
(1000, 2)


Creates an elfen extraction pipeline.
Specifies:
Dataset
Language
Text column

In [4]:
#Step 3: Initialize the Extractor
extractor = elfen.Extractor( 
    data=df, 
    language="en", 
    text_column="text" 
    ) 
print("Extractor created successfully!")

Extractor created successfully!


Extracts:

Type-Token Ratio (TTR)

Formula:
TTR = Unique Words / Total Words

Interpretation:
Higher TTR → Rich vocabulary
Lower TTR → Repetitive vocabulary

In [6]:
#Step 4: Extract a Single Feature
extractor.extract("ttr")
print(extractor.data.head())

shape: (5, 6)
┌─────┬──────────────────────────────┬─────────────────────────────┬──────────┬─────────┬──────────┐
│ id  ┆ text                         ┆ nlp                         ┆ n_tokens ┆ n_types ┆ ttr      │
│ --- ┆ ---                          ┆ ---                         ┆ ---      ┆ ---     ┆ ---      │
│ i64 ┆ str                          ┆ object                      ┆ u16      ┆ u16     ┆ f64      │
╞═════╪══════════════════════════════╪═════════════════════════════╪══════════╪═════════╪══════════╡
│ 1   ┆ I didn't know what to wear   ┆ I didn't know what to wear  ┆ 14       ┆ 13      ┆ 0.928571 │
│     ┆ to …                         ┆ to …                        ┆          ┆         ┆          │
│ 2   ┆ There used to be a big       ┆ There used to be a big      ┆ 12       ┆ 12      ┆ 1.0      │
│     ┆ cherry …                     ┆ cherry …                    ┆          ┆         ┆          │
│ 3   ┆ Sometimes, books and movies  ┆ Sometimes, books and movies ┆ 14      

Features Generated:
Flesch Reading Ease
Flesch-Kincaid Grade
SMOG
Gunning Fog
ARI
CLI
LIX
RIX

Gunning Fog = 8.45 → Approximately 8th-grade reading level.
SMOG = 8.84 → Around 9 years of education needed.
Flesch Reading Ease = 70.2 → Fairly easy to read.
Flesch-Kincaid Grade = 8.1 → Suitable for middle/high school students.

In [10]:
#Step 6: Extract Readability Features
extractor.extract_feature_group("readability")
#Step 7: Inspect Readability Results
print(
    extractor.data.select([
        "text",
        "gunning_fog",
        "smog",
        "flesch_reading_ease",
        "flesch_kincaid_grade"
    ]).head()
)

shape: (5, 5)
┌────────────────────────────┬─────────────┬──────────┬─────────────────────┬──────────────────────┐
│ text                       ┆ gunning_fog ┆ smog     ┆ flesch_reading_ease ┆ flesch_kincaid_grade │
│ ---                        ┆ ---         ┆ ---      ┆ ---                 ┆ ---                  │
│ str                        ┆ f64         ┆ f64      ┆ f64                 ┆ f64                  │
╞════════════════════════════╪═════════════╪══════════╪═════════════════════╪══════════════════════╡
│ I didn't know what to wear ┆ 8.457143    ┆ 8.841846 ┆ -907.175            ┆ 143.27               │
│ to …                       ┆             ┆          ┆                     ┆                      │
│ There used to be a big     ┆ 4.8         ┆ 3.1291   ┆ -905.145            ┆ 142.49               │
│ cherry …                   ┆             ┆          ┆                     ┆                      │
│ Sometimes, books and       ┆ 5.6         ┆ 3.1291   ┆ -907.175            ┆

Surface Features:

Number of tokens
Number of sentences
Average word length

Dependency Features:

Grammatical relationships

Morphological Features:

Word structure information

In [12]:
#Step 8: Extract Additional Linguistic Dimensions
extractor.extract_feature_group("surface")
extractor.extract_feature_group("dependency")
extractor.extract_feature_group("morphology")

print(extractor.data.head())

Feature group morphology not found. Check spelling.
shape: (5, 74)
┌─────┬────────────┬────────────┬──────────┬───┬────────────┬────────────┬────────────┬────────────┐
│ id  ┆ text       ┆ nlp        ┆ n_tokens ┆ … ┆ n_dependen ┆ n_dependen ┆ n_dependen ┆ n_dependen │
│ --- ┆ ---        ┆ ---        ┆ ---      ┆   ┆ cy_quantmo ┆ cy_relcl   ┆ cy_root    ┆ cy_xcomp   │
│ i64 ┆ str        ┆ object     ┆ u16      ┆   ┆ d          ┆ ---        ┆ ---        ┆ ---        │
│     ┆            ┆            ┆          ┆   ┆ ---        ┆ u16        ┆ u16        ┆ u16        │
│     ┆            ┆            ┆          ┆   ┆ u16        ┆            ┆            ┆            │
╞═════╪════════════╪════════════╪══════════╪═══╪════════════╪════════════╪════════════╪════════════╡
│ 1   ┆ I didn't   ┆ I didn't   ┆ 14       ┆ … ┆ 0          ┆ 0          ┆ 0          ┆ 1          │
│     ┆ know what  ┆ know what  ┆          ┆   ┆            ┆            ┆            ┆            │
│     ┆ to wear to ┆ to 

In [13]:
#Step 9: Extract All Features
extractor.extract_features()
print(extractor.data.head())

Extracting raw_sequence_length...
Extracting n_tokens...
Extracting n_sentences...
Extracting n_tokens_per_sentence...
Extracting n_characters...
Extracting avg_word_length...
Extracting n_types...
Extracting n_long_words...
Extracting n_lemmas...
Extracting n_per_morph_feature...
Extracting tree_width...
Extracting tree_depth...
Extracting tree_branching...
Extracting n_noun_chunks...
Extracting n_per_dependency_type...
Extracting sentiment_score...


c:\Users\kanch\AppData\Local\Programs\Python\Python311\Lib\site-packages\elfen\resources.py:89: UserWarning: The resource for feature sentiment_nrc has to be downloaded manually due to website restrictions. Please see the instructions in the GitHub repository for more information.
  warnings.warn(f"The resource for feature {feature} has to be "


FileNotFoundError: The system cannot find the path specified. (os error 3): ..._resources\Emotion\Sentiment\NRC-Emotion-Lexicon\NRC-Emotion-Lexicon-Wordlevel-v0.92.txt (set POLARS_VERBOSE=1 to see full path)

elfen successfully extracted several feature groups, including:

Surface Features
Readability Features
Lexical Richness Features
Morphological Features
Dependency Features
Syntactic Features

However, when the framework attempted to extract emotion-based features, it required an external linguistic resource called the NRC Emotion Lexicon.

Error Encountered
FileNotFoundError:
NRC-Emotion-Lexicon-Wordlevel-v0.92.txt

Why This Happened

The NRC Emotion Lexicon is not distributed automatically with the elfen package due to licensing and website restrictions.

As a result, the file was not available in the expected resource directory, causing the extraction process to terminate.

Impact
Previously extracted features remained available.
Emotion and sentiment-related features could not be generated.
Full extraction of all 1,061 features was not completed.

Solution

To enable complete feature extraction:

Download the NRC Emotion Lexicon manually.
Place the file in the resource directory expected by elfen.
Re-run:
extractor.extract_features()

The extraction process successfully generated multiple linguistic feature groups. However, the execution stopped when elfen attempted to compute emotion-related features that depend on the NRC Emotion Lexicon. Since this external resource was not installed, a FileNotFoundError occurred. This highlights a common challenge in NLP research, where tools often rely on additional linguistic resources beyond the main software package.

major challenges during elfen implementation


Errors Encountered During elfen Installation & Import

1. NumPy Installation Error

**Error:**

```text id="0g4o2y"
metadata-generation-failed
numpy==1.26.4
```

**Cause:**

* Python 3.13 was incompatible with the required NumPy version.
* No C/C++ compiler was available to build NumPy from source.

**Solution:**

* Switched to Python 3.11.

---

 2. Module Not Found: click

**Error:**

```text id="4mhm8t"
ModuleNotFoundError: No module named 'click'
```

**Cause:**

* spaCy dependency was missing.

**Solution:**

```bash id="40p5dz"
pip install click
```

---

3. Module Not Found: spaCy Model

**Error:**

```text id="gvgmbi"
OSError: [E050] Can't find model 'en_core_web_sm'
```

**Cause:**

* The English language model required by spaCy was not installed.

**Solution:**

```bash id="u7c6bj"
python -m spacy download en_core_web_sm
```

---

 4. pip Command Executed Inside Python Shell

**Error:**

```text id="2snjlwm"
SyntaxError: invalid syntax
```

Example:

```python id="7s4p3w"
>>> pip install numpy
```

**Cause:**

* Installation commands were executed inside the Python interpreter instead of Command Prompt.

**Solution:**

* Run pip commands in Terminal or Command Prompt.


5. NRC Emotion Lexicon Missing

**Error:**

```text id="r0k6ck"
FileNotFoundError:
NRC-Emotion-Lexicon-Wordlevel-v0.92.txt
```

**Cause:**

* External emotion lexicon required by elfen was not installed.

**Impact:**

* Full extraction of all 1,061 features could not be completed.

